spring-vault-core/src/test/java/org/springframework/vault/core/ReactiveVaultKeyValueMetadataTemplateIntegrationTests.java

```diff
@@ -0,0 +1,156 @@
+	}
+
+	@BeforeEach
+	void setup() {
+
+		vaultKeyValueMetadataOperations = vaultOperations.opsForVersionedKeyValue("versioned").opsForKeyValueMetadata();
+
+		for (var key : List.of(SECRET_NAME, CAS_SECRET_NAME)) {
+			vaultKeyValueMetadataOperations.delete(key)
+				.onErrorResume(e -> Mono.empty())
+				.as(StepVerifier::create)
+				.verifyComplete();
+		}
+
+		var secret = new HashMap<>();
+		secret.put("key", "value");
+
+		kvOperations.put(SECRET_NAME, secret).as(StepVerifier::create).verifyComplete();
+	}
+
+	@Test
+	void shouldReadMetadataForANewKVEntry() {
+
+		vaultKeyValueMetadataOperations.get(SECRET_NAME).as(StepVerifier::create).assertNext(metadataResponse -> {
+			assertThat(metadataResponse.getMaxVersions()).isEqualTo(0);
+			assertThat(metadataResponse.getCurrentVersion()).isEqualTo(1);
+			assertThat(metadataResponse.getVersions()).hasSize(1);
+			assertThat(metadataResponse.isCasRequired()).isFalse();
+			assertThat(metadataResponse.getCreatedTime().isBefore(Instant.now())).isTrue();
+			assertThat(metadataResponse.getUpdatedTime().isBefore(Instant.now())).isTrue();
+
+			var version1 = metadataResponse.getVersions().get(0);
+
+			if (prepare().getVersion().isGreaterThanOrEqualTo(Version.parse("1.2.0"))) {
+
+				assertThat(metadataResponse.getDeleteVersionAfter()).isEqualTo(Duration.ZERO);
+
+				assertThat(version1.getDeletedAt()).isNull();
+				assertThat(version1.getCreatedAt()).isBefore(Instant.now());
+			}
+
+			assertThat(version1.getVersion().getVersion()).isEqualTo(1);
+
+		}).verifyComplete();
+	}
+
+	@Test
+	void shouldUpdateMetadataVersions() {
+
+		Map<String, Object> secret = Map.of("newkey", "newvalue");
+		kvOperations.put(SECRET_NAME, secret).as(StepVerifier::create).verifyComplete();
+
+		vaultKeyValueMetadataOperations.get(SECRET_NAME).as(StepVerifier::create).assertNext(metadataResponse -> {
+			assertThat(metadataResponse.getCurrentVersion()).isEqualTo(2);
+			assertThat(metadataResponse.getVersions()).hasSize(2);
+		}).verifyComplete();
+
+	}
+
+	@Test
+	void shouldUpdateKVMetadata() {
+
+		var secret = Map.of("key", "value");
+
+		kvOperations.put(CAS_SECRET_NAME, secret).as(StepVerifier::create).verifyComplete();
+
+		Duration duration = Duration.ofMinutes(30).plusHours(6).plusSeconds(30);
+		VaultMetadataRequest request = VaultMetadataRequest.builder()
+			.casRequired(true)
+			.deleteVersionAfter(duration)
+			.maxVersions(20)
+			.build();
+
+		vaultKeyValueMetadataOperations.put(CAS_SECRET_NAME, request).as(StepVerifier::create).verifyComplete();
+
+		vaultKeyValueMetadataOperations.get(CAS_SECRET_NAME)
+			.as(StepVerifier::create)
+			.assertNext(metadataResponseAfterUpdate -> {
+				assertThat(metadataResponseAfterUpdate.isCasRequired()).isEqualTo(request.isCasRequired());
+				assertThat(metadataResponseAfterUpdate.getMaxVersions()).isEqualTo(request.getMaxVersions());
+
+				if (prepare().getVersion().isGreaterThanOrEqualTo(Version.parse("1.2.0"))) {
```

Comment:
Nit: We should obtain the version outside of the reactive flow to avoid potential blocking issues.

Reply:
Will update.